Passive acquire mode

TaranDahl opened the pull request:
### Passive acquire mode

- Now you can order the units to enter "Ceasefire Mode" or "Aggressive Mode", just like in RA3.
  - In ceasefire mode, units will only attack the targets specified by the player.
  - In aggressive mode, units will automatically attack all enemies, including buildings without weapons.
  - In normal mode, units will auto-target like usual.
- You can use [two](User-Interface#ceasefire-mode) [hotkeys](User-Interface#aggressive-mode) to switch between these two modes if  `[General] -> EnablePassiveAcquireMode` is set to `true`. When you press one of these hotkeys:
  - If all selected units are in that mode, they will switch to the normal mode. Otherwise, they will enter that mode. Units that cannot switch modes are not counted.
  - If any unit successfully switches its mode, a sound effect will be emitted. This sound can be customized through `[TechnoType] -> Voice(Enter/Exit)(Aggressive/Ceasefire)Mode`.
- You can use `[TechnoType] -> PassiveAcquireMode.Togglable` to specify whether the unit can toggle its mode.
- You can use `[TechnoType] -> PassiveAcquireMode` to specify the unit's initial mode.

In `rulesmd.ini`:
```ini
[General]
EnablePassiveAcquireMode=false        ; boolean

[SOMETECHNO]                          ; TechnoType
PassiveAcquireMode=Normal             ; passive acquire mode, Normal / Aggressive / Ceasefire
PassiveAcquireMode.Togglable=true     ; boolean
VoiceEnterAggressiveMode=             ; Sound entry, default to VoiceAttack or VoiceMove
VoiceExitAggressiveMode=              ; Sound entry, default to VoiceMove or VoiceSelect
VoiceEnterCeasefireMode=              ; Sound entry, default to VoiceSelect or VoiceMove
VoiceExitCeasefireMode=               ; Sound entry, default to VoiceAttack or VoiceMove
```

### `[ ]` Ceasefire Mode

- Order the selected units to enter or exit the ceasefire mode. See [this](New-or-Enhanced-Logics#passive-acquire-mode) for details.
- For localization add `TXT_CEASEFIRE_MODE`, `TXT_CEASEFIRE_MODE_DESC`, `MSG:CEASEFIRE_MODE_ON` and `MSG:CEASEFIRE_MODE_OFF` into your `.csf` file.

### `[ ]` Aggressive Mode

- Order the selected units to enter or exit the aggressive mode. See [this](New-or-Enhanced-Logics#passive-acquire-mode) for details.
- For localization add `TXT_AGGRESSIVE_MODE`, `TXT_AGGRESSIVE_MODE_DESC`, `MSG:AGGRESSIVE_MODE_ON` and `MSG:AGGRESSIVE_MODE_OFF` into your `.csf` file.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added Passive Acquire Mode with Normal, Aggressive, and Ceasefire states.
  * Added hotkeys to toggle Aggressive and Ceasefire modes for eligible selected units.
  * Aggressive units can automatically acquire targets, while Ceasefire units avoid initiating attacks and retaliation.
  * Added per-unit configuration for default mode, toggling, voice feedback, and building exemptions.

* **Documentation**
  * Added configuration and hotkey guidance for Passive Acquire Mode.
  * Updated the changelog and credits.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TaranDahl committed:
impl

TaranDahl committed:
Update Hooks.TargetEvaluation.cpp

TaranDahl committed:
AggressiveMode

TaranDahl committed:
Update Commands.cpp

TaranDahl committed:
Update Phobos.vcxproj

TaranDahl committed:
update

TaranDahl committed:
docs and update

Metadorius commented:
Previously when discussing stances we figured there should be a button/indicator at bottom bar. Have this went anywhere? If I recall correctly, someone even had the code to add custom buttons there.

Coronia commented:
Could there be visual indicator on techno to tell which of them is in aggressive/passive stance if it’s toggleable?

TaranDahl commented:
Maybe it could be drawn next to the unit like the insignia. 
But I don't think this is a good idea. We already have too many things to draw next to the unit. 
I'm asking Crim to open a pr for his bottom bar.

TaranDahl committed:
Update Hooks.cpp

Metadorius commented:
> @Metadorius The existing code has exceeded 500 lines. I think it's better to complete this PR before initiating a new one to finish these tasks, to avoid overwhelming the reviewer with too much code. What do you think?

@TaranDahl I propose to take CrimRecya's implementation of the bottom bar buttons from #1453 separately as a new PR without any extra logics (maybe could do a new button for triggering Select Next Idle Harvester command; however, I am not sure if the buttons have to be always tied to CommandClass, this is up for discussion), reviewing/refactoring it so it fits our standards, merging, and then rebasing this PR so it can use it.

CrimRecya commented on src/Commands/PassiveAcquireMode.cpp:
```diff
@@ -0,0 +1,210 @@
+				isAllSelectedUnitCeasefireMode = false;
+				TechnoVectorNonCeasefire.push_back(pTechno);
+			}
+			return;
```
Same.

CrimRecya commented on src/Ext/Rules/Body.h:
```diff
@@ -1,4 +1,4 @@
-#pragma once
+﻿#pragma once
```
Same as above.

CrimRecya commented on src/Ext/TechnoType/Body.h:
```diff
@@ -1,4 +1,4 @@
-#pragma once
+﻿#pragma once
```
Same as above.

CrimRecya commented on src/Ext/Techno/Body.cpp:
```diff
@@ -802,6 +802,98 @@ void TechnoExt::HandleOnDeployAmmoChange(TechnoClass* pThis, int maxAmmoOverride
+
+	const auto pThis = this->OwnerObject();
+	const auto pTechnoType = this->TypeExtData->OwnerObject();
+	int voiceIndex;
```
Sorry, it's just a suggestion. I just don't like having an uninitialized variable. You can try encapsulating the following code into a static inline or lambda function called `process...` and returning a voice index.

CrimRecya reviewed (approved):
There are some minor issues, but overall it looks good. I think the bottom bar part can be added after merging, of course, it depends on what you think.

TaranDahl commented:
> @TaranDahl I propose to take CrimRecya's implementation of the bottom bar buttons from https://github.com/Phobos-developers/Phobos/pull/1453 separately as a new PR without any extra logics (maybe could do a new button for triggering Select Next Idle Harvester command; however, I am not sure if the buttons have to be always tied to CommandClass, this is up for discussion), reviewing/refactoring it so it fits our standards, merging, and then rebasing this PR so it can use it.

@CrimRecya How about this

TaranDahl committed:
core

impl

Update Hooks.TargetEvaluation.cpp

AggressiveMode

Update Commands.cpp

Update Phobos.vcxproj

update

docs and update

Update Hooks.cpp

TaranDahl commented on src/Commands/Commands.cpp:
```diff
@@ -12,6 +12,14 @@
+#include <MouseClass.h>
+#include <WWMouseClass.h>
+
+#include <Utilities/Macro.h>
```
done.

TaranDahl committed:
Update Hooks.TargetEvaluation.cpp

TaranDahl committed:
Update Hooks.TargetEvaluation.cpp

TaranDahl committed:
update

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -376,3 +389,23 @@ Action __fastcall InfantryClass__WhatAction_Wrapper(InfantryClass* pThis, void*
 #pragma endregion
+
+#pragma region PassiveAcquireMode
+
+DEFINE_HOOK(0x6F8E1F, TechnoClass_SelectAutoTarget_CeasefireMode, 0x6)
+{
+	GET(TechnoTypeClass*, pType, EAX);
+	GET(TechnoClass*, pThis, ESI);
+	R->CL(pType->NoAutoFire || (TechnoExt::ExtMap.Find(pThis)->GetPassiveAcquireMode()) == PassiveAcquireMode::Ceasefire);
+	return R->Origin() + 0x6;
+}
```
Changed.
